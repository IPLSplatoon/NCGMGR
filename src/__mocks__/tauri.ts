export const mockTauriEvent = {
    listen: jest.fn()
}

jest.mock('@tauri-apps/api/event', () => mockTauriEvent)

export const mockTauri = {
    invoke: jest.fn()
}

jest.mock('@tauri-apps/api/tauri', () => mockTauri)

export const mockTauriDialog = {
    open: jest.fn()
}

jest.mock('@tauri-apps/api/dialog', () => mockTauriDialog)

export const mockTauriFs = {
    readDir: jest.fn(),
    readTextFile: jest.fn(),
    removeDir: jest.fn(),
    removeFile: jest.fn(),
    createDir: jest.fn(),
    writeFile: jest.fn()
}

jest.mock('@tauri-apps/api/fs', () => mockTauriFs)

export const mockTauriOs = {
    type: jest.fn()
}

jest.mock('@tauri-apps/api/os', () => mockTauriOs)

export const mockTauriShell = {
    open: jest.fn()
}

jest.mock('@tauri-apps/api/shell', () => mockTauriShell)
