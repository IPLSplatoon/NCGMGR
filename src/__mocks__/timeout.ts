export function mockTimeout (): void {
    jest.spyOn(window, 'setTimeout').mockImplementation(handler => {
        // noinspection SuspiciousTypeOfGuard
        if (typeof handler !== 'string') {
            handler()
        }
        return 0
    })
}
