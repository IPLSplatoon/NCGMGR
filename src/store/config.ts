import { createStore, Store, useStore } from 'vuex'
import { InjectionKey } from 'vue'

export const LOCAL_STORAGE_CONFIG_KEY = 'config'

export interface Configuration {
    installPath: string
}

export const configStore = createStore<Configuration>({
    state: {
        installPath: ''
    },
    mutations: {
        setInstallPath: (state, installPath: string) => {
            state.installPath = installPath
        },
        setState: (state, config: Configuration) => {
            Object.assign(state, config)
        }
    },
    actions: {
        load: store => {
            const config = JSON.parse(localStorage.getItem(LOCAL_STORAGE_CONFIG_KEY) ?? '{}')
            store.commit('setState', {
                ...store.state,
                ...config
            })
        },
        persist: store => {
            localStorage.setItem(LOCAL_STORAGE_CONFIG_KEY, JSON.stringify(store.state))
        }
    }
})

export const configStoreKey: InjectionKey<Store<Configuration>> = Symbol()

export function useConfigStore (): Store<Configuration> {
    return useStore(configStoreKey)
}
