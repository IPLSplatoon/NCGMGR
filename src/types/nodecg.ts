export interface NodecgConfiguration {
    port: number
}
