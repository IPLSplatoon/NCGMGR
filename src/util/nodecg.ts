import npa from 'npm-package-arg'

type HostedGit = { git: () => string }

export function normalizeBundlePath (path: string): { isValid: boolean, bundleName?: string, bundleUrl?: string } {
    try {
        const parsedPath = npa(path)
        if (!parsedPath.hosted) {
            return { isValid: false }
        }

        const gitUrl = (parsedPath.hosted as unknown as HostedGit).git()
        if (!gitUrl) {
            return { isValid: false }
        }

        const temp = gitUrl.split('/').pop()

        return {
            isValid: true,
            bundleName: temp?.substr(0, temp.length - 4),
            bundleUrl: gitUrl?.replace('git://', 'https://')
        }
    } catch (e) {
        return { isValid: false }
    }
}
