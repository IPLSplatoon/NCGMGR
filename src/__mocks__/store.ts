import { createStore, Store } from 'vuex'
import { Configuration } from '@/store/config'
import { LogStore } from '@/store/log'

export function createConfigStore (): Store<Configuration> {
    return createStore<Configuration>({
        state: {
            installPath: '/install/path'
        },
        actions: {
            load: jest.fn(),
            persist: jest.fn()
        },
        mutations: {
            setInstallPath: jest.fn()
        }
    })
}

export function createLogStore (): Store<LogStore> {
    return createStore<LogStore>({
        state: {
            lines: [],
            completed: false
        },
        mutations: {
            reset: jest.fn(),
            insertLine: jest.fn(),
            setCompleted: jest.fn()
        }
    })
}
