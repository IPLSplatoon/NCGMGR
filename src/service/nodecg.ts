import { readDir, readTextFile, removeDir, removeFile } from '@tauri-apps/api/fs'
import { PackageSchema } from '@/types/package'
import isEmpty from 'lodash/isEmpty'
import { InstallStatus } from '@/store/nodecg'
import { invoke } from '@tauri-apps/api/tauri'

export async function getNodecgStatus (directory: string): Promise<{ status: InstallStatus, message: string }> {
    if (isEmpty(directory?.trim())) {
        return {
            status: InstallStatus.UNABLE_TO_INSTALL,
            message: 'Please select an installation directory.'
        }
    }

    const dir = await readDir(directory)
    if (dir.length < 1) {
        return {
            status: InstallStatus.READY_TO_INSTALL,
            message: 'Directory is empty. Ready to install...'
        }
    } else {
        const packageFile = dir.find(entry => entry.name === 'package.json')
        if (packageFile) {
            const packageJson: PackageSchema = JSON.parse(await readTextFile(packageFile.path))
            if (packageJson.name === 'nodecg') {
                return {
                    status: InstallStatus.INSTALLED,
                    message: `Found NodeCG v${packageJson.version}.`
                }
            } else {
                return {
                    status: InstallStatus.UNABLE_TO_INSTALL,
                    message: `Found unknown package "${packageJson.name}".`
                }
            }
        } else {
            return {
                status: InstallStatus.UNABLE_TO_INSTALL,
                message: 'Could not find package.json.'
            }
        }
    }
}

export interface Bundle {
    name: string
    version: string
}

export async function getBundles (directory: string): Promise<Bundle[]> {
    if (isEmpty(directory.trim())) {
        throw new Error('No bundle directory provided.')
    }

    const bundlesDir = await readDir(directory + '/bundles', { recursive: true })
    return Promise.all(
        bundlesDir
            .filter(entry => entry.children?.some(child => child.name === 'package.json'))
            .map(async (dir) => {
                const packageFile = dir.children?.find(child => child.name === 'package.json')?.path
                if (!packageFile) {
                    throw new Error(`Missing package.json in directory ${dir.name}`)
                }

                const packageJson = JSON.parse(await readTextFile(packageFile)) as PackageSchema

                return {
                    name: packageJson.name,
                    version: packageJson.version
                }
            }))
}

export async function getBundleVersions (bundleName: string, nodecgPath: string): Promise<string[]> {
    return invoke('fetch_bundle_versions', { bundleName, nodecgPath })
}

export async function configFileExists (bundleName: string, nodecgPath: string): Promise<boolean> {
    const configDir = await readDir(`${nodecgPath}/cfg`)
    return configDir.some(item => item.name === `${bundleName}.json`)
}

export async function removeBundle (bundleName: string, nodecgPath: string): Promise<[void, void]> {
    return Promise.all([
        removeDir(`${nodecgPath}/bundles/${bundleName}`, { recursive: true }),
        (async () => {
            if (await configFileExists(bundleName, nodecgPath)) {
                return removeFile(`${nodecgPath}/cfg/${bundleName}.json`)
            }
        })()
    ])
}
