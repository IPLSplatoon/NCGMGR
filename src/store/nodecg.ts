import { createStore, Store, useStore } from 'vuex'
import { Bundle, getBundles, getNodecgStatus } from '@/service/nodecg'
import { configStore } from '@/store/config'
import { InjectionKey } from 'vue'

export enum NodecgStatus {
    UNKNOWN,
    READY_TO_INSTALL,
    INSTALLED,
    UNABLE_TO_INSTALL
}

export interface NodecgStore {
    status: {
        status: NodecgStatus
        message: string
        bundlesLoading: boolean
    },
    bundles: Bundle[]
}

export const nodecgStore: Store<NodecgStore> = createStore<NodecgStore>({
    state: {
        status: {
            status: NodecgStatus.UNKNOWN,
            message: '',
            bundlesLoading: false
        },
        bundles: []
    },
    actions: {
        async checkNodecgStatus (store) {
            store.state.status.status = NodecgStatus.UNKNOWN
            store.state.status.message = 'Checking status...'

            try {
                const { status, message } = await getNodecgStatus(configStore.state.installPath)
                store.state.status.message = message
                store.state.status.status = status
                if (status === NodecgStatus.INSTALLED) {
                    store.dispatch('getBundleList')
                } else {
                    store.state.bundles = []
                }
            } catch (e) {
                store.state.status.message = e.toString()
                store.state.status.status = NodecgStatus.UNABLE_TO_INSTALL
                store.state.bundles = []
            }
        },
        async getBundleList (store) {
            store.state.status.bundlesLoading = true
            store.state.bundles = await getBundles(configStore.state.installPath)
            store.state.status.bundlesLoading = false
        }
    }
})

export const nodecgStoreKey: InjectionKey<Store<NodecgStore>> = Symbol()

export function useNodecgStore (): Store<NodecgStore> {
    return useStore(nodecgStoreKey)
}
