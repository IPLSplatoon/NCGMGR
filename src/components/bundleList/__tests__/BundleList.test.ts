import { useNodecgStore } from '@/store/nodecg'
import { config, flushPromises, mount } from '@vue/test-utils'
import BundleSettings from '@/components/bundleList/BundleList.vue'
import { createTestingPinia, TestingPinia } from '@pinia/testing'
import { useConfigStore } from '@/store/config'
import { mockTauri } from '@/__mocks__/tauri'
import { IplButton } from '@iplsplatoon/vue-components'
import { removeBundle } from '@/service/nodecg'
import Mock = jest.Mock

jest.mock('@/service/nodecg')

describe('BundleSettings', () => {
    let pinia: TestingPinia
    config.global.stubs = {
        IplButton: true
    }

    beforeEach(() => {
        pinia = createTestingPinia()
        config.global.plugins = [pinia]
    })

    it('matches snapshot when bundles are loading', () => {
        const nodecgStore = useNodecgStore()
        nodecgStore.status.bundlesLoading = true
        const wrapper = mount(BundleSettings)

        expect(wrapper.html()).toMatchSnapshot()
    })

    it('matches snapshot when no bundles are present', () => {
        const nodecgStore = useNodecgStore()
        nodecgStore.bundles = []
        const wrapper = mount(BundleSettings)

        expect(wrapper.html()).toMatchSnapshot()
    })

    it('handles uninstalling a bundle', async () => {
        const configStore = useConfigStore()
        configStore.installPath = '/nodecg/path'
        const nodecgStore = useNodecgStore()
        nodecgStore.bundles = [
            { name: 'Bundle One', version: '1.2.3' },
            { name: 'Bundle Two', version: '5.0' }
        ]
        nodecgStore.getBundleList = jest.fn()
        const wrapper = mount(BundleSettings);
        (removeBundle as Mock).mockResolvedValue({})
        const bundleRow = wrapper.get('[data-test="bundle_Bundle One"]')

        bundleRow.getComponent<typeof IplButton>('[data-test="uninstall-button"]').vm.$emit('click')
        await flushPromises()

        expect(wrapper.findComponent('[data-test="uninstall-overlay"]').exists()).toEqual(true)
        expect(mockTauri.invoke).not.toHaveBeenCalled()

        wrapper.getComponent<typeof IplButton>('[data-test="confirm-uninstall-button"]').vm.$emit('click')
        await flushPromises()

        expect(wrapper.findComponent('[data-test="uninstall-overlay"]').exists()).toEqual(false)
        expect(removeBundle).toHaveBeenCalledWith('Bundle One', '/nodecg/path')
        expect(nodecgStore.getBundleList).toHaveBeenCalled()
    })

    it('handles bundle uninstallation being cancelled', async () => {
        const nodecgStore = useNodecgStore()
        nodecgStore.bundles = [
            { name: 'Bundle One', version: '1.2.3' },
            { name: 'Bundle Two', version: '5.0' }
        ]
        const wrapper = mount(BundleSettings)
        mockTauri.invoke.mockResolvedValue({})
        const bundleRow = wrapper.get('[data-test="bundle_Bundle One"]')

        bundleRow.getComponent<typeof IplButton>('[data-test="uninstall-button"]').vm.$emit('click')
        await flushPromises()

        expect(wrapper.findComponent('[data-test="uninstall-overlay"]').exists()).toEqual(true)

        wrapper.getComponent<typeof IplButton>('[data-test="cancel-uninstall-button"]').vm.$emit('click')
        await flushPromises()

        expect(wrapper.findComponent('[data-test="uninstall-overlay"]').exists()).toEqual(false)
        expect(mockTauri.invoke).not.toHaveBeenCalled()
    })
})
