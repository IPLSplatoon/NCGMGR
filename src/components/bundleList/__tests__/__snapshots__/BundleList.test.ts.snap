// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BundleSettings matches snapshot when bundles are loading 1`] = `<div class="text-center m-b-8"> Loading... </div>`;

exports[`BundleSettings matches snapshot when no bundles are present 1`] = `<div class="text-center m-b-8"> No bundles installed. </div>`;
